### PR TITLE
fix refresh_token expiration policy per service

### PR DESCRIPTION
**Pull request by antoine777**
Currently, when we set a `refreshTokenExpirationPolicy` in an OAuth2 service, an `OAuth20RefreshTokenExpirationPolicy` is created regardless of the `removeDescendantTickets` setting (which is `false` by default).

The consequence is that `refresh_token` cannot be used when the associated TGT no longer exists (CAS considers `refresh_token` to be expired).

The fix creates the correct ExpirationPolicy based on the `removeDescendantTickets` setting.

Related to [PR 5619](https://github.com/apereo/cas/pull/5619) but with more complete tests

### CODE REVIEW

File: support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/ticket/refreshtoken/OAuth20DefaultRefreshTokenFactory.java [hunk 1]
```diff
@@ -45,9 +45,15 @@ public class OAuth20DefaultRefreshTokenFactory implements OAuth20RefreshTokenFac
      */
     protected final ServicesManager servicesManager;
 
+    /**
+     * Indicates whether RefreshToken issued and linked to a ticket-granting ticket
+     * should also be removed as part of logout
+     */
+    protected final  boolean removeDescendantTickets;
```
Comment (mmoayyed) [inline support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/ticket/refreshtoken/OAuth20DefaultRefreshTokenFactory.java:52]: Please run checkstyle on your changeset.

File: support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/ticket/refreshtoken/OAuth20DefaultRefreshTokenFactory.java [hunk 2]
```diff
@@ -45,9 +45,15 @@ public class OAuth20DefaultRefreshTokenFactory implements OAuth20RefreshTokenFac
      */
     protected final ServicesManager servicesManager;
 
+    /**
```
Comment (mmoayyed) [inline support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/ticket/refreshtoken/OAuth20DefaultRefreshTokenFactory.java:48]: This comment is incorrect; this is not strictly what this setting does.

Your comment would also break javadocs.

File: support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/ticket/refreshtoken/OAuth20DefaultRefreshTokenFactory.java [hunk 3]
```diff
@@ -45,9 +45,15 @@ public class OAuth20DefaultRefreshTokenFactory implements OAuth20RefreshTokenFac
...
+     * Indicates whether RefreshToken issued and linked to a ticket-granting ticket
```
Comment (mmoayyed) [inline support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/ticket/refreshtoken/OAuth20DefaultRefreshTokenFactory.java:49]: Classes pointed out in javadocs should be using proper javadocs tags.

File: support/cas-server-support-oauth/src/test/java/org/apereo/cas/ticket/refreshtoken/OAuth20DefaultRefreshTokenFactorySovereignExpirationPolicyTests.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+import java.util.Map;
+import java.util.Set;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+/**
+ * This is {@link OAuth20DefaultRefreshTokenFactorySovereignExpirationPolicyTests}.
+ *
+ * @since 6.2.0
```
Comment (mmoayyed) [inline support/cas-server-support-oauth/src/test/java/org/apereo/cas/ticket/refreshtoken/OAuth20DefaultRefreshTokenFactorySovereignExpirationPolicyTests.java:24]: Wrong version.

File: support/cas-server-support-oauth/src/test/java/org/apereo/cas/ticket/refreshtoken/OAuth20DefaultRefreshTokenFactorySovereignExpirationPolicyTests.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ */
+@Tag("OAuthToken")
+@TestPropertySource(properties = "cas.logout.remove-descendant-tickets=false")
```
Comment (mmoayyed) [inline support/cas-server-support-oauth/src/test/java/org/apereo/cas/ticket/refreshtoken/OAuth20DefaultRefreshTokenFactorySovereignExpirationPolicyTests.java:27]: You will need to make sure your tests run with or without this setting

**Review by egglestonahma — APPROVED** [review]

**Pull request closed without merging**